aten/src/ATen/ParallelOpenMP.cpp

```diff
@@ -0,0 +1,80 @@
+#ifdef AT_PARALLEL_OPENMP
+#include <ATen/Parallel.h>
+
+#include <atomic>
+
+#ifdef TH_BLAS_MKL
+#include <mkl.h>
+#endif
+
+namespace at {
+
+namespace {
+// Number of threads set by the user
+std::atomic<int> num_threads{-1};
+
+} // namespace
+
+void init_num_threads() {
+  auto nthreads = num_threads.load();
+  if (nthreads > 0) {
+    set_num_threads(nthreads);
+  } else {
+#if defined(_OPENMP) && defined(TH_BLAS_MKL)
+  // If we are using MKL an OpenMP make sure the number of threads match.
+  // Otherwise, MKL and our OpenMP-enabled functions will keep changing the
+  // size of the OpenMP thread pool, resulting in worse performance (and memory
+  // leaks in GCC 5.4)
+  omp_set_num_threads(mkl_get_max_threads());
```

Comment:
What about the OMP_NUM_THREADS environment variable? A lot of users define it. Will that still work? What if MKL isn't available?

Reply:
I did the analysis of how it works in https://github.com/pytorch/pytorch/issues/19001